### PR TITLE
fix(bootstrap): no longer creates KMS master key by default

### LINKED ISSUE #10115: CDKToolkit stack with `newStyleStackSynthesis` creates CMK that costs 1USD/month
Please make this behaviour opt-in, so that by default, the CDKToolkit does not cost anything. I never expected a support stack to cost money. There is already a standard S3 KMS key that costs nothing, and it's enough for trying things out.

### PR DESCRIPTION
**Pull request by rix0rrr**
The modern bootstrap stack used to unconditionally create a KMS Customer
Master Key (CMK) for users. This incurs a $1/month charge for every user
of the CDK for every region and account they want to deploy in, which is
not acceptable if we're going to make this the default bootstrapping
experience in the future.

This PR switches off the creation of the CMK by default for new
bootstrap stacks. Bootstrap stacks that already exist can remove the
existing CMK by running:

```
cdk bootstrap --bootstrap-customer-key=false [aws://...]
```

This change is backwards compatible: updates to existing (modern)
bootstrap stacks will leave the current KMS key in place. To achieve
this, the new default is encoded into the CLI, not into the template.

Fixes #10115.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by hoegertn** [comment]
Are there any downsides of using the default S3 key? Is there any cross-account access needed or policies that are not possible with AWS managed CMKs?

**Comment by rix0rrr** [comment]
> Are there any downsides of using the default S3 key? Is there any cross-account access needed or policies that are not possible with AWS managed CMKs?

Not for this key, no. For the keys that are created as part of the Pipeline, yes. I'm working on a separate PR there.

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-environment.ts [hunk 1]
```diff
@@ -139,3 +158,8 @@ export class Bootstrapper {
     }
   }
 }
+
+/**
+ * Magic parameter value that will cause the bootstrap-template.yml to NOT create a CMK but use the default keyo
+ */
+const USE_AWS_MANAGED_KEY = '-';
```
Comment (njlynch) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-environment.ts:165]: `~` Any reason not to make this a little less magical? Like 
```suggestion
const USE_AWS_MANAGED_KEY = 'USE_AWS_MANAGED_KEY';
```

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-environment.ts [hunk 2]
```diff
@@ -93,15 +100,27 @@ export class Bootstrapper {
...
+    // * If an ARN is given, that ARN. Otherwise:
+    //   * '-' if customerKey = false
+    //   * '' if customerKey = true
+    //   * if customerKey is also not given
+    //     * undefined if we already had a value in place (reusing what we had)
+    //     * '-' if this is the first time we're deploying this stack (or upgrading from old to new bootstrap)
+    const kmsKeyId = params.kmsKeyId ??
+      (params.createCustomerMasterKey === true ? '' :
+        params.createCustomerMasterKey === false ? USE_AWS_MANAGED_KEY :
+          current.parameters.FileAssetsBucketKmsKeyId !== undefined ? undefined : USE_AWS_MANAGED_KEY);
```
Comment (njlynch) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-environment.ts:113]: `~` Take or leave it, but removed a ternary, defined a new constant, and got this:

```suggestion
    const currentKmsKeyId = current.parameters.FileAssetsBucketKmsKeyId;
    const kmsKeyId = params.kmsKeyId ??
      (params.createCustomerMasterKey ? CREATE_CMK
        : (params.createCustomerMasterKey === false || currentKmsKeyId === undefined)
          ? USE_AWS_MANAGED_KEY : undefined);
```

**Review by njlynch — APPROVED** [review]
Looks good to me; two nitpicks.

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-environment.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-environment.ts:165]: I guess not!

**Comment by rix0rrr** [comment]
Tested to work with x-account and x-account/x-region pipeline

**Pull request merged**